modules/peridynamics/src/mesh/FileMeshPD.C

```diff
@@ -0,0 +1,220 @@
+  pd_mesh.reserve_nodes(_total_nodes);
+  pd_mesh.reserve_elem(_total_bonds);
+
+  // loop through all pd_nodes to generate PD mesh nodes structure
+  for (unsigned int i = 0; i < _total_nodes; ++i)
+    pd_mesh.add_point(_pdnode[i].coord, i);
+
+  // generate PD mesh
+  unsigned int k = 0;
+  for (unsigned int i = 0; i < _total_nodes; ++i)
+    for (unsigned int j = 0; j < _node_neighbors[i].size(); ++j)
+      if (_node_neighbors[i][j] > i)
+      {
+        Elem * pd_elem = pd_mesh.add_elem(new Edge2);
+        pd_elem->set_id() = k;
+        pd_elem->set_node(0) = pd_mesh.node_ptr(i);
+        pd_elem->set_node(1) = pd_mesh.node_ptr(_node_neighbors[i][j]);
+        // block id for pd mesh from fe mesh
+        pd_elem->subdomain_id() = _pdnode[i].blockID;
+        ++k;
+      }
+
+  // convert boundary info from fe_boundary_info to pd_boundary_info
+  // build element list for user specified nodal boundaries
+  std::vector<dof_id_type> elems;
+  std::vector<unsigned short int> sides;
+  std::vector<boundary_id_type> ids;
+
+  BoundaryInfo & fe_boundary_info = _fe_mesh->get_boundary_info();
+  fe_boundary_info.build_side_list_from_node_list();
+  fe_boundary_info.build_active_side_list(elems, sides, ids);
+
+  unsigned int n = elems.size();
+  // array of boundary elems
+  std::vector<BndElement *> bnd_elems(n);
+  // map of set of elem IDs connected to each boundary
+  std::map<boundary_id_type, std::set<dof_id_type>> bnd_elem_ids;
+  for (unsigned int i = 0; i < n; ++i)
+  {
+    bnd_elems[i] = new BndElement(_fe_mesh->elem_ptr(elems[i]), sides[i], ids[i]);
+    bnd_elem_ids[ids[i]].insert(elems[i]);
+  }
+  // for nodeset only
+  std::set<boundary_id_type> fe_node_bid(fe_boundary_info.get_node_boundary_ids());
+  std::set<boundary_id_type> fe_edge_bid(fe_boundary_info.get_edge_boundary_ids());
+  if (!fe_edge_bid.empty())
+    mooseError("FileMeshPD currently only accepts nodesets!");
+
+  for (std::set<boundary_id_type>::iterator bit = fe_node_bid.begin(); bit != fe_node_bid.end();
+       ++bit)
+  {
+    pd_boundary_info.nodeset_name(*bit) = fe_boundary_info.get_nodeset_name(*bit);
+    for (MeshBase::element_iterator eit = _fe_mesh->elements_begin();
+         eit != _fe_mesh->elements_end();
+         ++eit)
+    {
+      Elem * fe_elem = *eit;
+      std::map<boundary_id_type, std::set<dof_id_type>>::const_iterator it =
+          bnd_elem_ids.find(*bit);
+      if (it != bnd_elem_ids.end())
+        if (it->second.find(fe_elem->id()) != it->second.end())
+          pd_boundary_info.add_node(pd_mesh.node_ptr(fe_elem->id()), *bit);
+    }
+  }
+
+  // define center and right nodes, ONLY for geometry of circular cross section
+  // centered at the origin
+  for (unsigned int i = 0; i < _total_nodes; ++i)
+  {
+    Real X = (_pdnode[i].coord)(0);
+    Real Y = (_pdnode[i].coord)(1);
+    Real Z = (_pdnode[i].coord)(2);
+    Real dis = std::sqrt(X * X + Y * Y + Z * Z);
+    if (dis < 0.001)
```

Comment:
@hchen139 Instead of hard coding in a length parameter like 0.001 here, you think it would be better to make it a function of horizon (1/3 or 1/2 horizon or something like that) ?

Reply:
You are right, but we need to pre calculate a quantity such as average horizon size for the irregular mesh. Sometimes this may be problematic since a global average horizon size may not necessarily presenting the local horizon size accurately for some irregular meshes.

Reply:
I think we can calculate the minimum spacing between two material points for a irregular discretization, and use this value to assign boundary ids. Will do this.